packages/npm/send/android/src/main/java/com/candlefinance/send/SendModule.kt

```diff
@@ -1,25 +1,173 @@
 package com.candlefinance.send
 
+import com.facebook.react.bridge.Promise
 import com.facebook.react.bridge.ReactApplicationContext
 import com.facebook.react.bridge.ReactContextBaseJavaModule
 import com.facebook.react.bridge.ReactMethod
-import com.facebook.react.bridge.Promise
+import kotlinx.serialization.Serializable
+import kotlinx.serialization.SerializationException
+import kotlinx.serialization.encodeToString
+import kotlinx.serialization.json.Json
+import okhttp3.HttpUrl
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okhttp3.RequestBody.Companion.toRequestBody
+import okhttp3.internal.http.HttpMethod
+import okio.Buffer
+import okio.ByteString.Companion.decodeBase64
+import okio.GzipSource
+import okio.IOException
+import java.util.concurrent.TimeUnit
+
+// FIXME: Actually check charset parameter first if set, then fall back to MIME type
+private fun bodyIsUTF8(contentTypeHeader: String?, utf8ContentTypes: List<String>): Boolean {
+  return contentTypeHeader != null && utf8ContentTypes.contains(contentTypeHeader.split(';')[0])
+}
+
+@Serializable
+data class Request(
+  val baseURL: String,
+  val path: String,
+  val queryParameters: Map<String, String>,
+  val headerParameters: Map<String, String>,
+  val method: String,
+  val body: String?,
+  val utf8ContentTypes: List<String>,
+) {
+  val httpUrl: HttpUrl
+    get() {
+      val urlBuilder = (baseURL + path).toHttpUrl().newBuilder()
+      queryParameters.forEach { (key, value) ->
+        urlBuilder.addQueryParameter(key, value)
+      }
+
+      return urlBuilder.build()
+    }
+
+  val okHttpRequest: okhttp3.Request
+    get() {
+      val requestBuilder = okhttp3.Request.Builder().url(httpUrl)
+      headerParameters.forEach { (key, value) ->
+        requestBuilder.addHeader(key, value)
+      }
+
+      requestBuilder.method(
+        method,
+        if (body == null) if (HttpMethod.requiresRequestBody(method)) "".toRequestBody() else null
+        else {
```

Comment:
As is, this took me a minute to parse..
```suggestion
        if (body == null) 
            if (HttpMethod.requiresRequestBody(method)) "".toRequestBody() else null
        else {
```